### PR TITLE
feat: add canonical handling

### PR DESCRIPTION
**Pull request by thelostone-mc**
#### When creating a linked Profile and the canonical already exists

```
{
  projects(
    filter: {id: {equalTo: "0xbfe075166257fda5b915e3dda8be67b44b2db34ddf476841b1e7aa33a5e13db2"}}
  ) {
    id
    chainId
    projectType
    metadata
  }
}
```

image

#### When attempting to create a linked Profile BUT the canonical does not exist

image

Fixes https://github.com/gitcoinco/grants-stack-indexer/issues/427

### CODE REVIEW

File: src/indexer/allo/v2/handleEvent.ts [hunk 1]
```diff
@@ -19,6 +20,20 @@ function generateRoundRoles(poolId: bigint) {
...
+  const canonicalProject = (metadata as Record<string, unknown>)["canonical"] as CanonicalProject;
+
+  if (!canonicalProject) {
+    return null;
+  }
+
+  const canonicalRegistryAddress = canonicalProject.registryAddress;
+  const canonicalChainId = canonicalProject.chainId;
+
+  // TODO: fetch the canonical project from the database
```
Comment (thelostone-mc) [inline src/indexer/allo/v2/handleEvent.ts:33]: Do we have an example I can use as a reference ?

File: src/indexer/allo/v2/handleEvent.ts [hunk 2]
```diff
@@ -25,6 +25,21 @@ function generateRoundRoles(poolId: bigint) {
...
 
+async function getProjectType(db: Database, profileId: Hex, metadata: JSON) {
+
+  if (metadata.hasOwnProperty("canonical")) {
+    // Linked Profile
+    const canonical: CanonicalProject = (metadata as any)["canonical"] as CanonicalProject;
+    const project = await db.getProjectById(canonical.chainId, profileId);
+    if (!project) {
+      throw new Error("Project not found");
+    }
```
Comment (gravityblast) [inline src/indexer/allo/v2/handleEvent.ts:36]: I suggest not to throw an error here. The project might not exist yet if the other chain is slower to index

Comment (gravityblast) [reply src/indexer/allo/v2/handleEvent.ts:36]: I would rather set the `type` based on the metadata without checking the db

**Review by boudra — CHANGES_REQUESTED** [review]
I noticed some castings and some issues that will be errors for the linter
I want to suggest this as a more robust way to do the classification (i didn't try the code so it might need modification):

```ts
const linkedProjectMetadata = z.object({
    "canonical": z.object({
           chainId: z.number(),
           registryAddress: z.string()
    })
});

if(linkedProjectMetadata.safeParse(metadata).success) {
   return "linked"
}

return "canonical";
}
```
without the CanonicalProject type.

this way we can start validating the metadatas

File: src/indexer/allo/v2/handleEvent.ts [hunk 3]
```diff
@@ -25,6 +26,21 @@ function generateRoundRoles(poolId: bigint) {
   return { managerRole, adminRole };
 }
 
+function getProjectType(metadata: object) {
+  const linkedProjectMetadata = z.object({
+    canonical: z.object({
+      chainId: z.string(),
```
Comment (thelostone-mc) [inline src/indexer/allo/v2/handleEvent.ts:32]:
```suggestion
      chainId: z.number(),
```

**Review by boudra — APPROVED** [review]
⚠️ before you merge please increase the schema version or the schema changes won't be applied: https://github.com/gitcoinco/grants-stack-indexer?tab=readme-ov-file#deployment ⚠️

**Pull request merged**